load_tests/load_test.py

```diff
@@ -0,0 +1,169 @@
+
+# With multiple codebuild projects running parallel,
+# Testing resources only needs to be created once
+def create_testing_resources():
+    if OUTPUT_PLUGIN != 'firehose':
+        client = boto3.client('cloudformation')
+        waiter = client.get_waiter('stack_exists')
+        waiter.wait(
+            StackName=TESTING_RESOURCES_STACK_NAME,
+            WaiterConfig={
+                'MaxAttempts': 60
+            }
+        )
+        waiter = client.get_waiter('stack_create_complete')
+        waiter.wait(
+            StackName=TESTING_RESOURCES_STACK_NAME
+        )
+    else:
+        os.chdir(f'./load_tests/{sys.argv[1]}')
+        os.system('cdk deploy --require-approval never')
+
+# For tests on ECS, we need to:
+#  1. generate and register task definitions based on templates at /load_tests/task_definitons
+#  2. run tasks with different throughput levels for 10 mins
+#  3. wait until tasks completed, set buffer for logs sent to corresponding destinations
+#  4. validate logs and print the result
+def run_ecs_tests():
+    ecs_cluster_name = os.environ['ECS_CLUSTER_NAME']
+    client = boto3.client('ecs')
+    waiter = client.get_waiter('tasks_stopped')
+    processes = set()
+    names = locals()
+
+    # Run ecs tasks and store task arns
+    for throughput in THROUGHPUT_LIST:
+        os.environ['THROUGHPUT'] = throughput
+        generate_task_definition(throughput)
+        response = client.run_task(
+                cluster=ecs_cluster_name,
+                launchType='EC2',
+                taskDefinition=f'{PREFIX}{OUTPUT_PLUGIN}-{throughput}'
+        )
+        names[f'{OUTPUT_PLUGIN}_{throughput}_task_arn'] = response['tasks'][0]['taskArn']
+
+    # Wait until task stops and start validation
+    for throughput in THROUGHPUT_LIST:
+        waiter.wait(
+            cluster=ecs_cluster_name,
+            tasks=[
+                names[f'{OUTPUT_PLUGIN}_{throughput}_task_arn'],
+            ],
+            WaiterConfig={
+                'MaxAttempts': 600
+            }
+        )
+        response = client.describe_tasks(
+            cluster=ecs_cluster_name,
+            tasks=[
+                names[f'{OUTPUT_PLUGIN}_{throughput}_task_arn'],
+            ]
+        )
+        check_exit_code(response)
+        input_record = calculate_total_input_number(throughput)
+        log_delay = get_log_delay(response)
+        # Buffer for waiting all logs sent to destinations
+        time.sleep(180)
+        # Validate logs
+        if OUTPUT_PLUGIN == 'cloudwatch':
+            os.environ['LOG_PREFIX'] = throughput
+            os.environ['DESTINATION'] = OUTPUT_PLUGIN
+        else:
+            os.environ['LOG_PREFIX'] = f'{OUTPUT_PLUGIN}-test/ecs/{throughput}/'
+            os.environ['DESTINATION'] = 's3'
+        processes.add(subprocess.Popen(['go', 'run', './load_tests/validation/validate.go', input_record, log_delay]))
```

Comment:
So validate downloads all content and stores in list. Then processes list. Content up to 30 mb/s * 600s  * 4validators = 72,000mb or 72GB. I'm concerned that the RAM may be over utilized causing issues. What do you think?

Reply:
So they are not in one instance. Each output plugin has its own instance (codebuild project) to run so it shouldn't be the case.

Reply:
Followed up offline. Confirm vm has enough mem for 72/4 = 18GB, and pipeline works.